Sources_v3/Workers/Background/MissingEventsReplayer.swift

```diff
@@ -0,0 +1,100 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+/// The type is designed to obtain missing events that happened in watched channels while user
+/// was not connected to the web-socket.
+///
+/// The object listenes for `ConnectionStatusUpdated` events
+/// and remembers the `CurrentUserDTO.lastReceivedEventDate` when status becomes `connecting`.
+///
+/// When the status becomes `connected` the `/sync` endpoint is called
+/// with `lastReceivedEventDate` and `cids` of watched channels.
+///
+/// We remember `lastReceivedEventDate` when state becomes `connecting` to catch the last event date
+/// before the `HealthCheck` override the `lastReceivedEventDate` with the recent date.
+///
+class MissingEventsReplayer<ExtraData: ExtraDataTypes>: Worker {
+    // MARK: - Properties
+    
+    private var connectionObserver: EventObserver?
+    private var lastSyncedAt: Date?
+    
+    // MARK: - Init
+
+    override init(database: DatabaseContainer, webSocketClient: WebSocketClient, apiClient: APIClient) {
+        super.init(database: database, webSocketClient: webSocketClient, apiClient: apiClient)
+        startObserving()
+    }
+    
+    // MARK: - Private
+    
+    private func startObserving() {
+        connectionObserver = EventObserver(
+            notificationCenter: webSocketClient.eventNotificationCenter,
+            tranform: { $0 as? ConnectionStatusUpdated },
+            callback: { [unowned self] in
+                switch $0.connectionStatus {
+                case .connecting:
+                    self.obtainLastSyncDate()
+                case .connected:
+                    self.fetchAndReplayMissingEvents()
+                default:
+                    break
+                }
+            }
+        )
+    }
+    
+    private func obtainLastSyncDate() {
+        lastSyncedAt = database.backgroundReadOnlyContext.currentUser()?.lastReceivedEventDate
+    }
+    
+    private func fetchAndReplayMissingEvents() {
+        guard let lastSyncedAt = lastSyncedAt else { return }
+        
+        let watchedChannelIDs = channels.filter { $0.isWatched }.map(\.cid)
+        
+        let endpoint: Endpoint<MissingEventsPayload<ExtraData>> = .missingEvents(
+            since: lastSyncedAt,
+            cids: watchedChannelIDs
+        )
+        
+        apiClient.request(endpoint: endpoint) { [weak self] in
+            switch $0 {
+            case let .success(payload):
+                self?.webSocketClient.eventNotificationCenter.process(payload.eventPayloads)
```

Comment:
Yeah, this is not ideal. Accessing `eventNotificationCenter` via `webSocketClient` but I think it's OK for now.